include/revng/AutoEnforcer/AutoEnforcerContext.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <initializer_list>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Casting.h"
+
+#include "revng/AutoEnforcer/KindsRegistry.h"
+#include "revng/AutoEnforcer/SavableObject.h"
+
+namespace AutoEnforcer {
+class NamedGlobal {
+public:
+  NamedGlobal(llvm::StringRef Name, SavableObjectBase &Obj) :
+    Name(Name), Obj(&Obj) {}
+
+  llvm::StringRef getName() const { return Name; }
+
+  SavableObjectBase &getObj() const { return *Obj; }
+
+private:
+  llvm::StringRef Name;
+  SavableObjectBase *Obj;
+};
+
+class Context {
+public:
+  explicit Context(KindsRegisty registry) : KRegistry(std::move(registry)) {}
+  Context();
+  Context(llvm::ArrayRef<NamedGlobal> Globals);
+  Context(const Context &) = delete;
+  Context(Context &&) = delete;
+  Context &operator=(const Context &) = delete;
+  Context &operator=(Context &&) = delete;
+  ~Context() = default;
+
+  template<typename T>
+  llvm::Expected<T *> get(llvm::StringRef Name) const {
+    auto Iter = AEGlobals.find(Name);
+    if (Iter == AEGlobals.end())
+      return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                     "pipeline loader context did not "
+                                     "contained object %s",
+                                     Name.str().c_str());
+    auto *Casted = llvm::dyn_cast<T>(Iter->second);
+    if (Casted == nullptr)
+      return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                     "pipeline loader was requested to cast %s "
+                                     "to the wrong type",
+                                     Name.str().c_str());
+
+    return Casted;
+  }
+
+  void add(llvm::StringRef Name, SavableObjectBase &Object) {
+    AEGlobals.try_emplace(Name, &Object);
+  }
+
+  const KindsRegisty &getKindRegistry() const { return KRegistry; }
+
+  llvm::Error save(llvm::StringRef Path) const;
+  llvm::Error load(llvm::StringRef Path);
```

Comment:
Terminology: we have `SavableObjectBase`, `save`, `load`. What if we go for something on the same line but using the termine *(de)serializable*?
Or maybe just call it *global*?

Reply:
mh deserialize kind of entaills that it's turned into a string, while save and load store and load from disk

Reply:
Mh, can we take a stream then? Sounds more general. Or would that be a problem when you spawn external commands?
"Savable" is quite odd.

Reply:
a stream is not ok because theoretically one global object may wish to save more than one file. The path is the path to the dir/globalname, and they are allowed to append postfixes to create more files. 

i can change savable to something better yeah

Reply:
Are you sure we want more than one file? What's the use case?

Reply:
it's not probably the case of globals object but as an example, if for some reason we ends up in a situations in which a enforcer wishes to compile independently isolated function, then the containers is allowed to create a directory at the provided path and store the various  object file independently.

i don't think this is very common toh, furthermore this allows globals and containers to raise appropriate errors when the file is in a unexpected state, (worng permissions, does/does not exits). Or allows to use third parties functions that expects a path instead of a stream to perform some operations (not likelly as well)

Reply:
> if for some reason we ends up in a situations in which a enforcer wishes to compile independently isolated function...

Uhm, we're talking about globals, which do not have a granularity and are not the output of an enforcer.

In general, let's try to limit the degrees of freedom we have to plausible use cases. The more constrained we are, the less complexity (no need to handle not fully foreseen scenarios) and the less opportunity for error (at the cost of having to make changes in the future).

Anyway, the only upside of a stream is if we ever want to serialize the thing in a string (`.dump()`?).
But other than that, we can stick with saving to disk.